src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs

```diff
@@ -0,0 +1,91 @@
+
+        //This is the OSGroup folder that we want to select test zips from. This does not require an exact match as some tests 
+        //may or may not have an exact OSGroup folder match.
+        [Required]
+        public string FilterToOSGroup { get; set; }
+        //This is the path under a given Platform/OSGroup/Configuration folder where the Test Archives are placed
+        [Required]
+        public string TestArchivesRoot { get; set; }
+
+        [Output]
+        public ITaskItem[] FilteredUploadItems
+        {
+            get { return this._copyFilteredUploadItems.ToArray(); }
+        }
+
+        private readonly List<ITaskItem> _copyFilteredUploadItems = new List<ITaskItem>();
+
+        public override bool Execute()
+        {
+            string[] filesToSearch = Directory.GetFiles(TestOutputDirListsLocation);
+            
+            foreach (var filename in filesToSearch)
+            {
+                string selectedTest = string.Empty;
+                using (StreamReader sr = new StreamReader(File.OpenRead(filename)))
+                {
+                    string line;
+                    while ((line = sr.ReadLine()) != null)
+                    {
+                        //Each line is of the format "$(BinDir)/tests/AnyOS.AnyCPU.Debug/Tests/netcoreapp1.0"
+                        //Each test zip is of the format "$(BinDir)/tests/AnyOS.AnyCPU.Debug/archive/tests/netcoreapp1.0/Tests.zip"
+                        //This is an implicit contract maintained in corefx 
+                        if (calculateMatchValue(line) > calculateMatchValue(selectedTest))
+                        {
+                            int lastIndexOfSlash = line.LastIndexOf('/');
+                            string parentDirectory = line.Substring(0, lastIndexOfSlash);
+                            // This is to split "$(BinDir)/tests/AnyOS.AnyCPU.Debug/Tests" and "/netcoreapp1.0"
+                            string testTFMFolderInArchive = line.Substring(lastIndexOfSlash + 1);
+                            lastIndexOfSlash = parentDirectory.LastIndexOf('/');
+                            // This is to split "$(BinDir)/tests/AnyOS.AnyCPU.Debug/" and "Tests"
+                            string testArchiveDir = Path.Combine(
+                                parentDirectory.Substring(0, lastIndexOfSlash),
+                                TestArchivesRoot);
+                            //Another contract in corefx is that the test name should be the AssemblyName appended with .Tests.
+                            string testZipName = parentDirectory.Substring(lastIndexOfSlash + 1) + ".zip";
+                            selectedTest = Path.Combine(testArchiveDir, testTFMFolderInArchive, testZipName);
+                        }
+                    }
+                }
+                if (!string.IsNullOrEmpty(selectedTest))
+                {
+                    this._copyFilteredUploadItems.Add(new TaskItem(selectedTest));
```

Comment:
I think "this" is generally not used in CoreFX/BuildTools unless it's necessary (which it isn't here)